src/test/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloaderTest.java

```diff
@@ -0,0 +1,178 @@
+
+    Path downloaderDestination = downloader.call();
+    Assert.assertEquals(destination, downloaderDestination);
+  }
+
+  @Test
+  public void testDownloadURL_worksWithNullProgressListener()
+      throws IOException, InterruptedException {
+
+    Path destination = tmp.getRoot().toPath().resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(1);
+    URL fakeRemoteResource = testSourceFile.toUri().toURL();
+
+    StandardDownloader downloader =
+        new StandardDownloader(fakeRemoteResource, destination, null, null);
+
+    Path downloaderDestination = downloader.call();
+    Assert.assertEquals(destination, downloaderDestination);
+  }
+
+  @Test
+  public void testDownloadURL() throws IOException, InterruptedException {
+    Path destination = tmp.getRoot().toPath().resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(StandardDownloader.BUFFER_SIZE * 10 + 1);
+    URL fakeRemoteResource = testSourceFile.toUri().toURL();
+    DownloadProgressListener mockListener = Mockito.mock(DownloadProgressListener.class);
+
+    StandardDownloader downloader =
+        new StandardDownloader(fakeRemoteResource, destination, null, mockListener);
+
+    Path downloaderDestination = downloader.call();
+    Assert.assertEquals(destination, downloaderDestination);
+
+    Assert.assertTrue(FileUtils.contentEquals(destination.toFile(), testSourceFile.toFile()));
+
+    ArgumentCaptor<Long> lastChunkCaptor = ArgumentCaptor.forClass(Long.class);
+    ArgumentCaptor<Long> completedCaptor = ArgumentCaptor.forClass(Long.class);
+    ArgumentCaptor<Long> totalCaptor = ArgumentCaptor.forClass(Long.class);
+    Mockito.verify(mockListener, Mockito.atLeastOnce())
+        .updateProgress(
+            lastChunkCaptor.capture(), completedCaptor.capture(), totalCaptor.capture());
+
+    // check we always present the same and correct file size
+    for (Long total : totalCaptor.getAllValues()) {
+      Assert.assertEquals(testSourceFile.toFile().length(), total.longValue());
+    }
+
+    // check out chunks add up to total
+    long sum = 0;
+    for (Long chunk : lastChunkCaptor.getAllValues()) {
+      sum += chunk;
+    }
+    Assert.assertEquals(testSourceFile.toFile().length(), sum);
+
+    // check completed matches chunk data
+    long last = 0;
+    for (int i = 0; i < completedCaptor.getAllValues().size(); i++) {
+      Assert.assertEquals(
+          last + lastChunkCaptor.getAllValues().get(i),
+          completedCaptor.getAllValues().get(i).longValue());
+      last = completedCaptor.getAllValues().get(i);
+    }
+  }
+
+  private Path createTestRemoteResource(long sizeInBytes) throws IOException {
+
+    Path testFile = tmp.newFile().toPath();
+    try (BufferedWriter writer = Files.newBufferedWriter(testFile, Charset.defaultCharset())) {
+      for (long i = 0; i < sizeInBytes; i++) {
+        writer.write('a');
+      }
+    }
+    return testFile;
```

Comment:
flush the writer

Reply:
Writer flushes when closing : https://docs.oracle.com/javase/7/docs/api/java/io/Writer.html#close()